src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -62,44 +62,62 @@ AsyncRPCOperation_sendmany::AsyncRPCOperation_sendmany(
 
     sendFromAccount_ = pwalletMain->FindAccountForSelector(ztxoSelector_).value_or(ZCASH_LEGACY_ACCOUNT);
 
-    // we always allow shielded change when not sending from the legacy account
-    if (sendFromAccount_ != ZCASH_LEGACY_ACCOUNT) {
-        allowedChangeTypes_.insert(libzcash::ChangeType::Sapling);
-    }
-
-    // calculate the target totals
+    // Determine the target totals and recipient pools
     for (const SendManyRecipient& recipient : recipients_) {
         std::visit(match {
             [&](const CKeyID& addr) {
                 transparentRecipients_ += 1;
                 txOutputAmounts_.t_outputs_total += recipient.amount;
-                allowedChangeTypes_.insert(libzcash::ChangeType::Transparent);
+                recipientPools_.insert(OutputPool::Transparent);
             },
             [&](const CScriptID& addr) {
                 transparentRecipients_ += 1;
                 txOutputAmounts_.t_outputs_total += recipient.amount;
-                allowedChangeTypes_.insert(libzcash::ChangeType::Transparent);
+                recipientPools_.insert(OutputPool::Transparent);
             },
             [&](const libzcash::SaplingPaymentAddress& addr) {
                 txOutputAmounts_.sapling_outputs_total += recipient.amount;
-                if (ztxoSelector_.SelectsSprout() && !allowRevealedAmounts_) {
-                    throw JSONRPCError(
+                recipientPools_.insert(OutputPool::Sapling);
+                if (!(ztxoSelector_.SelectsSapling() || allowRevealedAmounts_)) {
+                    if (ztxoSelector_.SelectsSprout()) {
```

Comment:
This approach looks as though it takes O(k<sup>2</sup>) code for k pools, but I guess we can address that when we add another pool.